pkg/cmd/agent-task/shared/log.go

```diff
@@ -0,0 +1,570 @@
+	lines := slices.DeleteFunc(strings.Split(string(logs), "\n"), func(line string) bool {
+		return line == ""
+	})
+
+	for _, line := range lines {
+		raw, found := strings.CutPrefix(line, "data: ")
+		if !found {
+			return false, errors.New("unexpected log format")
+		}
+
+		// The only log entry type we're interested in is a chat completion chunk,
+		// which can be verified by a successful unmarshal into the corresponding
+		// type AND the Object field being equal to "chat.completion.chunk". The
+		// latter is to avoid accepting an empty JSON object (i.e. "{}"). Also,
+		// if the entry is not what we expect, we should just skip and avoid
+		// returning an error.
+		var entry chatCompletionChunkEntry
+		err := json.Unmarshal([]byte(raw), &entry)
+		if err != nil || entry.Object != "chat.completion.chunk" {
+			continue
+		}
+
+		if stop, err := renderLogEntry(entry, w, io); err != nil {
+			return false, fmt.Errorf("failed to process log entry: %w", err)
+		} else if stop {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func renderLogEntry(entry chatCompletionChunkEntry, w io.Writer, io *iostreams.IOStreams) (bool, error) {
+	cs := io.ColorScheme()
+	var stop bool
+	for _, choice := range entry.Choices {
+		if choice.FinishReason == "stop" {
+			stop = true
+		}
+
+		if len(choice.Delta.ToolCalls) == 0 {
+			if choice.Delta.Content != "" && choice.Delta.Role == "assistant" {
+				// Copilot message and we should display.
+				renderRawMarkdown(choice.Delta.Content, w, io)
+			}
+			continue
+		}
+
+		// Since we don't want to clear-and-reprint live progress of events, we
+		// need to only process entries that correspond to a finished tool call.
+		// Such entries have a non-empty Content field.
+		if choice.Delta.Content == "" {
+			continue
+		}
+
+		if choice.Delta.ReasoningText != "" {
+			// Note that this should be formatted as a normal "thought" message,
+			// without the heading.
+			renderRawMarkdown(choice.Delta.ReasoningText, w, io)
+		}
+
+		for _, tc := range choice.Delta.ToolCalls {
+			name := tc.Function.Name
+			if name == "" {
+				continue
+			}
+
+			args := tc.Function.Arguments
+
+			switch name {
+			case "run_setup":
+				if v := unmarshal[runSetupToolArgs](args); v != nil {
+					renderToolCallTitle(w, cs, v.Name, "")
+					continue
+				}
+			case "view":
+				args := viewToolArgs{}
+				if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
+					return false, fmt.Errorf("failed to parse 'view' tool call arguments: %w", err)
+				}
+				fmt.Fprintf(w, "View %s\n", cs.Bold(relativeFilePath(args.Path)))
+
+				content := stripDiffFormat(choice.Delta.Content)
+
+				// TODO: Strip the diff formatting from this, but for now render as it is.
+				if err := renderFileContentAsMarkdown(args.Path, content, w, io); err != nil {
+					return false, fmt.Errorf("failed to render viewed file content: %w", err)
+				}
```

Comment:
I'm not sure if it's okay to bail out if the Markdown renderer fails, because I think it's quite possible for AI to produce flawed Markdown. In such cases, I think we should just ignore the error and continue processing logs.

One more reason is, if we bail out at this point, then the user would never be able to see their session logs, because again we reach to the same point and abort.